io/include/detray/io/common/detector_writer.hpp

```diff
@@ -0,0 +1,122 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s)
+#include "detray/io/common/detail/detector_component_writers.hpp"
+#include "detray/io/common/detail/type_traits.hpp"
+#include "detray/io/common/writer_types.hpp"
+
+// System include(s)
+#include <ios>
+
+namespace detray {
+
+namespace io {
+
+enum class format { json = 0u };
+
+}  // namespace io
+
+/// @brief Writer class for detray detectors.
+///
+/// Based on both the given config/file format and the detector type,
+/// the correct writers are being assembled and called.
+class detector_writer {
+
+    struct config {
+        /// Replace files in case they already exists
+        bool m_replace = false;
+        /// Compactify json output, if not json format this flag does nothing
+        bool m_compact_io = false;
+        /// Whether to write the material to file
+        bool m_write_material = true;
+        /// Whether to write the accelerator grids to file
+        bool m_write_grids = true;
+
+        /// Getters
+        /// @{
+        bool replace_files() const { return m_replace; }
+        bool compactify_json() const { return m_compact_io; }
+        bool write_material() const { return m_write_material; }
+        bool write_grids() const { return m_write_grids; }
+        /// @}
+
+        /// Setters
+        /// @{
+        config& replace_files(bool flag) {
+            m_replace = flag;
+            return *this;
+        }
+        config& compactify_json(bool flag) {
+            m_compact_io = flag;
+            return *this;
+        }
+        config& write_material(bool flag) {
+            m_write_material = flag;
+            return *this;
+        }
+        config& write_grids(bool flag) {
+            m_write_grids = flag;
+            return *this;
+        }
+        /// @}
+    };
+
+    public:
+    /// @returns access to the io config
+    config& get_config() { return m_config; }
+
+    /// Write @param det to file in the format @param format,
+    /// using @param names to name the components
+    template <class detector_t>
+    void write(detector_t& det, const typename detector_t::name_map& names,
```

Comment:
I can put this into a function, np, but it will not save us from the config state

Reply:
Done